primitives/src/pcs/mod.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Polynomial Commitment Scheme
+pub mod errors;
+mod multilinear_kzg;
+pub mod prelude;
+mod structs;
+mod transcript;
+mod univariate_kzg;
+
+use ark_ec::PairingEngine;
+use ark_serialize::CanonicalSerialize;
+use ark_std::{rand::RngCore, vec::Vec};
+use errors::PCSError;
+
+/// This trait defines APIs for polynomial commitment schemes.
+/// Note that for our usage of PCS, we do not require the hiding property.
+pub trait PolynomialCommitmentScheme<E: PairingEngine> {
+    /// Prover parameters
+    type ProverParam;
+    /// Verifier parameters
+    type VerifierParam;
+    /// Structured reference string
+    type SRS;
+    /// Polynomial and its associated types
+    type Polynomial;
+    /// Polynomial input domain
+    type Point;
+    /// Polynomial Evaluation
+    type Evaluation;
+    /// Commitments
+    type Commitment: CanonicalSerialize;
+    /// Batch commitments
+    type BatchCommitment: CanonicalSerialize;
+    /// Proofs
+    type Proof;
+    /// Batch proofs
+    type BatchProof;
+
+    /// Build SRS for testing.
+    ///
+    /// - For univariate polynomials, `supported_size` is the maximum degree.
+    /// - For multilinear polynomials, `supported_size` is the number of
+    ///   variables.
+    ///
+    /// WARNING: THIS FUNCTION IS FOR TESTING PURPOSE ONLY.
+    /// THE OUTPUT SRS SHOULD NOT BE USED IN PRODUCTION.
+    fn gen_srs_for_testing<R: RngCore>(
+        rng: &mut R,
+        supported_size: usize,
+    ) -> Result<Self::SRS, PCSError>;
+
+    /// Trim the universal parameters to specialize the public parameters.
+    /// Input both `supported_degree` for univariate and
+    /// `supported_num_vars` for multilinear.
+    fn trim(
+        srs: &Self::SRS,
+        supported_degree: usize,
+        supported_num_vars: Option<usize>,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), PCSError>;
```

Comment:
imo, input parameters that are potentially huge (exceed stack capacity of resource limited machine) should use type like: 

```rust
fn trim(srs: impl Borrow<Self::SRS>, ...);
fn commit(prover_param: impl Borrow<Self::ProverParam>, ...);
```

so this primarily referring to SRS and prover_param. other structs should be fine.
the syntax above allow for any types of borrow, from normal in-memory reference, to smart pointer, to atomic reference counters etc.

here's a simple illustration:
```rust
use std::borrow::Borrow;
use std::rc::Rc;
use std::sync::Arc;

fn test(_srs: impl Borrow<String>){}

fn main() {
    let srs = String::from("hi");
    let srs_boxed = Box::new(srs.clone());
    let srs_rc = Rc::new(srs.clone());
    let srs_arc = Arc::new(srs.clone());
    
    test(&srs);
    test(srs_boxed);
    test(srs_arc);
    test(srs_rc);
}
```

Reply:
One question: shall we make a similar change for other potentially large types? Like `Self::Polynomial`, `Self::BatchProof` etc?

Reply:
I'm neutral for `Self::Polynomial`. It can be large when the degree is high, possibly ~|SRS|/2 because it consists of field elements not group elements. On the other hand putting it in heaps might affect prover's efficiency.

For `Self::BatchProof` I'd say no because it won't be huge.

Reply:
Alright. I'll just update it for `SRS` and `ProverParam`.